scarb/src/core/registry/client/http.rs

```diff
@@ -250,3 +227,108 @@ impl HttpCacheKey {
         matches!(self, Self::None)
     }
 }
+
+impl<'c> IndexConfigManager<'c> {
+    fn new(source_id: SourceId, config: &'c Config) -> Self {
+        let cache_file_name = format!("{}.json", source_id.ident());
+        let cache_fs = config
+            .dirs()
+            .registry_dir()
+            .into_child("configs")
+            .into_child("http");
+
+        Self {
+            source_id,
+            config,
+            cache_file_name,
+            cache_fs,
+            cell: OnceCell::new(),
+        }
+    }
+
+    async fn load(&self) -> Result<&IndexConfig> {
+        self.cell
+            .get_or_try_init(|| self.load_impl_with_log())
+            .await
+            .context("failed to fetch registry config")
+    }
+
+    async fn load_impl_with_log(&self) -> Result<IndexConfig> {
+        let index_config = self.fetch().await?;
+        trace!(index_config = %serde_json::to_string(&index_config).unwrap());
+        Ok(index_config)
+    }
+
+    async fn fetch(&self) -> Result<IndexConfig> {
+        match self.fetch_from_cache().await {
+            Ok(Some(index_config)) => {
+                debug!("using cached config");
+                return Ok(index_config);
+            }
+            Ok(None) => {}
+            Err(err) => {
+                warn!("failed to fetch cached config: {err:?}");
+            }
+        }
+
+        let index_config = self.fetch_from_origin().await?;
+
+        if let Err(err) = self.save_in_cache(&index_config).await {
+            warn!("failed to save config in cache: {err:?}");
+        }
+
+        Ok(index_config)
+    }
+
+    fn may_be_cached(&self) -> bool {
+        self.cache_fs
+            .path_unchecked()
+            .join(&self.cache_file_name)
+            .exists()
+    }
+
+    async fn fetch_from_cache(&self) -> Result<Option<IndexConfig>> {
+        if !self.may_be_cached() {
+            return Ok(None);
+        }
+        let mut file = self
+            .cache_fs
+            .open_ro(&self.cache_file_name, &self.cache_file_name, self.config)?
+            .into_async();
+        let mut buffer = Vec::new();
+        file.read_to_end(&mut buffer).await?;
+        let index_config = serde_json::from_slice(&buffer)?;
+        Ok(Some(index_config))
+    }
+
+    async fn save_in_cache(&self, index_config: &IndexConfig) -> Result<()> {
+        let mut file = self
+            .cache_fs
+            .open_rw(&self.cache_file_name, &self.cache_file_name, self.config)?
+            .into_async();
+        let json = serde_json::to_vec_pretty(index_config)?;
```

Comment:
Why `to_vec_pretty`? 🤔

Reply:
Hmm, because I choose it arbitrarily 🤔 Will drop `pretty`, because it indeed isn't necessary.